### PR TITLE
[Form] add `choice_filter` option to `ChoiceType`

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #11847, #14072 |
| License | MIT |
| Doc PR | todo |
- [x] Add some tests
- [ ] Add more tests in `ChoiceTypeTest`
- [ ] Address a doc PR
- [ ] Update the CHANGELOG
- [ ] Add a note in the UPGRADE file

### CODE REVIEW

**Comment by HeahDude** [comment]
Basic usage:
```php
$builder->add('locales', LocaleType::class, array(
    'choice_filter' => function ($choice) use ($userLocales) {
        return in_array(key($choice), $userLocales);
    },
));

$builder->add('entities', EntityType::class, array(
    'class' => \AppBundle\Entity\Issue::class,
    'choice_filter' => function ($choice) use ($user) {
        return $user->canHandle($choice);
    },
));
```

**Comment by HeahDude** [comment]
Tests added.

**Comment by HeahDude** [comment]
This is a full BC implementation.

The only thing that should be noted in UPGRADE is that is someone uses its own `ChoiceListFactory` should either implement `FilteringChoiceListFactoryInterface` or use the `FilteredChoiceListDecorator` to use the new `choice_filter` option, otherwise it will simply be ignored.

File: src/Symfony/Component/Form/ChoiceList/Factory/FilteringFactoryDecorator.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createFilteredListFromLoader(ChoiceLoaderInterface $loader, $value = null, callable $filter)
+    {
+        // Don't hash the filter since the original choices may have been loaded already
+        // with a different filter if any.
+        $hash = CachingFactoryDecorator::generateHash(array($loader, $value));
```
Comment (HeahDude) [inline src/Symfony/Component/Form/ChoiceList/Factory/FilteringFactoryDecorator.php]: I'm just not sure about the over caching here, but it seems that considering this decorator might not decorate the `CachingFactoryDecorator`, keeping the resolved choices to filter should be the right thing to do.

**Comment by HeahDude** [comment]
@stof any thought on this one regarding the previous (wrong) implementation ?

**Comment by HeahDude** [comment]
ping @webmozart :)

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/FilteringFactoryDecoratorTest.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            ->with($filteredChoices)
+            ->willReturn('RESULT');
+
+        $this->assertSame('RESULT', $this->factory->createFilteredListFromLoader($choiceLoader, null, $filter));
+    }
+
+    /**
+     * @dataProvider provideChoicesAndFilter
+     */
+    public function testCreateFilteredChoiceListFromLoaderLoadsOriginalListOnFirstCall($choices, $choiceList, $filter, $filteredChoices)
```
Comment (trousers) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/FilteringFactoryDecoratorTest.php]: Codacy Issue found: [Avoid unused parameters such as '$choices'.](https://www.codacy.com/app/git_9/symfony/file/2332042926/issues/source?bid=3207821&fileBranchId=3207839#l84)

File: src/Symfony/Component/Form/ChoiceList/Factory/FilteringFactoryDecorator.php (hunk 1 above)
Comment (trousers) [inline src/Symfony/Component/Form/ChoiceList/Factory/FilteringFactoryDecorator.php]: Codacy Issue found: [Avoid using static access to class 'Symfony\Component\Form\ChoiceList\Factory\CachingFactoryDecorator' in method 'createFilteredListFromLoader'.](https://www.codacy.com/app/git_9/symfony/file/2332042754/issues/source?bid=3207821&fileBranchId=3207839#l93)

File: src/Symfony/Component/Form/Tests/ChoiceList/Factory/FilteringFactoryDecoratorTest.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            ->with($filteredChoices)
+            ->willReturn('RESULT');
+
+        $this->assertSame('RESULT', $this->factory->createFilteredListFromChoices($choices, null, $filter));
+    }
+
+    /**
+     * @dataProvider provideChoicesAndFilter
+     */
+    public function testCreateFilteredChoiceListFromLoader($choices, $choiceList, $filter, $filteredChoices)
```
Comment (trousers) [inline src/Symfony/Component/Form/Tests/ChoiceList/Factory/FilteringFactoryDecoratorTest.php]: Codacy Issue found: [Avoid unused parameters such as '$choices'.](https://www.codacy.com/app/git_9/symfony/file/2332042926/issues/source?bid=3207821&fileBranchId=3207839#l59)

File: src/Symfony/Component/Form/ChoiceList/Factory/FilteringFactoryDecorator.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Component/Form/ChoiceList/Factory/FilteringFactoryDecorator.php]: This caching is useless, as the decorated choice list already has caching

**Comment by stof** [comment]
@HeahDude this will have to wait until 3.2. We are already in the 3.1 feature freeze, and there is still too much to evaluate about this feature to merge it now (we are not yet fully freezed as we allow us to include some PRs which are ready but not yet merged, but this feature cannot be considered as ready yet IMO).
So I will focus on the review of other PRs for now (the ones still considered for inclusion in 3.1) and come back at this feature later.

**Comment by HeahDude** [comment]
@stof I absolutely agree ;) Thank you for your feedback!

Status: Needs Work

**Comment by ogizanagi** [comment]
@HeahDude : What about accepting an array in the `choice_filter` option?

The following:

```php
$builder->add('locales', LocaleType::class, array(
    'choice_filter' => $userLocales,
));
```

will automatically generate the callable using `in_array`.
It'll help when the type is configurable from a configuration file for instance (like in EasyAdminBundle).

**Comment by HeahDude** [comment]
>will automatically generate the callable using in_array.

I like the idea, thanks! I'll try to implement it soon when revisiting this one.

**Comment by HeahDude** [comment]
@ogizanagi Just curious why not simply use the `choices` option of the `ChoiceType` in this case?

**Comment by ogizanagi** [comment]
I expected this question 😅 

This allow to blindly filter the available values, without caring of the implementation of the type. Behind, the type has the responsibility to associate the proper labels for instance.

Using the `choices` option of the `ChoiceType` forces you to redefine everything.

**Comment by HeahDude** [comment]
I think you can already solve this case using a custom type:
```php
class UserLocaleType extends AbstractType
{
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver
            ->setRequired('display_locale')
            ->setNormalizer('choice_label', function (Options $options) {
                $displayLocale = $options['display_locale'];

                return function ($choice) use ($displayLocale) {
                    return Intl::getLocaleBundle()->getLocaleName($choice, $displayLocale);
                };
            })
        ;
    }

    public function getParent()
    {
        return ChoiceType::class;
    }
}

// Then you can do

```php
$builder->add('locales', UserLocaleType::class, array(
    'choices' => $userLocales,
    'display_locale' => $userLocale,
));
```

But I agree, the filter option definitely simplifies the implementation though :)

**Comment by HeahDude** [comment]
But regarding performances implementing a specific type like in my example may be better.

**Comment by ogizanagi** [comment]
I agree. Such uses of the `choice_filter` options will be reserved mainly for countries, locales, languages (...). I doubt it'll have a big impact on performances, but it has a very good one on DX.

**Comment by xabbuh** [comment]
Is this really something that will be commonly used? I mean if you need to filter your choices, why don't you just do that when specifying the `choices` option then?

**Comment by HeahDude** [comment]
@xabbuh in case of inheritance there would be no need to define again the `choices` and we could use the same cached choice list on which to apply the filter.

For the `EntityType` it would allow to filter `choices` thanks to dynamic properties or method that could not be used with a custom query builder.

So to me this feature is the right answer to both needs.

**Comment by xabbuh** [comment]
@HeahDude Sounds reasonable to me. Do you think you can finish this one?

**Comment by nicolas-grekas** [comment]
ping @HeahDude

**Comment by fabpot** [comment]
@HeahDude Do you think you can work on finishing this PR or shall we close it?

**Comment by fabpot** [comment]
Closing this PR as there is no more activity. @HeahDude Feel free to reopen whenever you have time to finish it. Thanks.

**Comment by yceruto** [comment]
See alternative https://github.com/symfony/symfony/pull/28607

**Pull request closed without merging**